pydantic_ai_slim/pydantic_ai/models/grok.py

```diff
@@ -0,0 +1,625 @@
+        run_context: RunContext[Any] | None = None,
+    ) -> AsyncIterator[StreamedResponse]:
+        """Make a streaming request to the Grok model."""
+        # Use injected client or create one in the current async context
+        client = self._client or AsyncClient(api_key=self._api_key)
+
+        # Convert messages to xAI format
+        xai_messages = self._map_messages(messages)
+
+        # Convert tools: combine built-in (server-side) tools and custom (client-side) tools
+        tools: list[chat_types.chat_pb2.Tool] = []
+        if model_request_parameters.builtin_tools:
+            tools.extend(self._get_builtin_tools(model_request_parameters))
+        if model_request_parameters.tool_defs:
+            tools.extend(self._map_tools(model_request_parameters))
+        tools_param = tools if tools else None
+
+        # Filter model settings to only include xAI SDK compatible parameters
+        xai_settings: dict[str, Any] = {}
+        if model_settings:
+            # Map pydantic_ai settings to xAI SDK parameters
+            if 'temperature' in model_settings:
+                xai_settings['temperature'] = model_settings['temperature']
+            if 'top_p' in model_settings:
+                xai_settings['top_p'] = model_settings['top_p']
+            if 'max_tokens' in model_settings:
+                xai_settings['max_tokens'] = model_settings['max_tokens']
+            if 'stop_sequences' in model_settings:
+                xai_settings['stop'] = model_settings['stop_sequences']
+            if 'seed' in model_settings:
+                xai_settings['seed'] = model_settings['seed']
+            if 'parallel_tool_calls' in model_settings:
+                xai_settings['parallel_tool_calls'] = model_settings['parallel_tool_calls']
+            if 'presence_penalty' in model_settings:
+                xai_settings['presence_penalty'] = model_settings['presence_penalty']
+            if 'frequency_penalty' in model_settings:
+                xai_settings['frequency_penalty'] = model_settings['frequency_penalty']
+
+        # Create chat instance
+        chat = client.chat.create(model=self._model_name, messages=xai_messages, tools=tools_param, **xai_settings)
+
+        # Stream the response
+        response_stream = chat.stream()
+        streamed_response = GrokStreamedResponse(
+            model_request_parameters=model_request_parameters,
+            _model_name=self._model_name,
+            _response=response_stream,
+            _timestamp=now_utc(),
+            _provider_name='xai',
+        )
+        yield streamed_response
+
+    def _process_response(self, response: chat_types.Response) -> ModelResponse:
+        """Convert xAI SDK response to pydantic_ai ModelResponse."""
+        from typing import cast
+
+        parts: list[ModelResponsePart] = []
+
+        # Add reasoning/thinking content first if present
+        if hasattr(response, 'reasoning_content') and response.reasoning_content:
```

Comment:
Why do we need the `hasattr`?

Reply:
Yes removed all the hasattr references for just direct property checks.